src/main/webapp/app/modules/provider/record/single-record-view.tsx

```diff
@@ -33,19 +61,68 @@ class SingleRecordView extends React.Component<ISingleRecordViewProps, ISingleRe
     locationWidths: [],
     isOrganizationOpen: false,
     isServicesOpen: false,
-    isLocationsOpen: false
+    isLocationsOpen: false,
+    currentServiceIdx: 0,
+    detailsView: false
   };
 
   toggleOrganization = () => this.setState({ isOrganizationOpen: !this.state.isOrganizationOpen });
-  toggleServices = () => this.setState({ isServicesOpen: !this.state.isServicesOpen });
+  toggleServices = () => this.setState({ isServicesOpen: !this.state.isServicesOpen, detailsView: false });
   toggleLocations = () => this.setState({ isLocationsOpen: !this.state.isLocationsOpen });
 
   componentDidMount() {
     this.props.getProviderEntity(this.props.match.params.orgId);
   }
 
+  componentWillUpdate(nextProps) {
+    if (nextProps.organization.id !== this.props.organization.id && nextProps.organization.locations && nextProps.organization.services) {
+      this.measureLocationsWidth(nextProps.organization);
+    }
+  }
+
+  showServiceDetails = serviceIdx => this.setState({ detailsView: true, currentServiceIdx: serviceIdx });
+
+  closeServiceDetails = () => this.setState({ detailsView: false });
+
+  nextService = servicesCount => {
+    const { currentServiceIdx } = this.state;
+    if (currentServiceIdx + 1 < servicesCount) {
+      this.setState({ currentServiceIdx: currentServiceIdx + 1 });
+    } else {
+      this.setState({ currentServiceIdx: 0 });
+    }
+  };
+
+  prevService = servicesCount => {
+    const { currentServiceIdx } = this.state;
+    if (currentServiceIdx === 0) {
+      this.setState({ currentServiceIdx: servicesCount - 1 });
```

Comment:
Is there a possibility when servicesCount == 0. That would cause currentServiceIdx to be negative and then "organization.services[currentServiceIdx].name" can cause NPE

Reply:
The part that uses this function is rendered only when the servicesCount > 0